flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/datastream/DataStreamTableFunctionScanRule.scala

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.rules.datastream
+
+import org.apache.calcite.plan.RelOptRule
+import org.apache.calcite.rel.RelNode
+import org.apache.calcite.rel.convert.ConverterRule
+import org.apache.flink.table.plan.nodes.FlinkConventions
+import org.apache.flink.table.plan.nodes.datastream.DataStreamTableFunctionScan
+import org.apache.flink.table.plan.nodes.logical.FlinkLogicalTableFunctionScan
+
+class DataStreamTableFunctionScanRule
+  extends ConverterRule(
+    classOf[FlinkLogicalTableFunctionScan],
+    FlinkConventions.LOGICAL,
+    FlinkConventions.DATASTREAM,
+    "DataStreamTableFunctionScanRule"
+  ) {
+
+  override def convert(rel: RelNode): RelNode = {
+    val scan: FlinkLogicalTableFunctionScan = rel.asInstanceOf[FlinkLogicalTableFunctionScan]
+    val traitSet = rel.getTraitSet.replace(FlinkConventions.DATASTREAM)
+
+    new DataStreamTableFunctionScan(
```

Comment:
Should this conversion happen always? How does it play along with for example `LogicalCorrelate` nodes, `DataStreamCorrelateRule` and `org.apache.flink.table.api.stream.sql.CorrelateTest`?

Does it work because if vulcano planer picks this rule in case of `LogicalCorrelate` it is later unable to convert `LogicalCorrelate` to `DataStreamCorrelate` and thus retracts from `DataStreamTableFunctionScanRule`?

Reply:
There might be two cases that TableFunction act like Source, one is 'Select * from Leteral Table(udtf())', the other is that the left table do not correlate with the right table so there's no
`LogicalCorrelate ` node.